sbtc-core/src/signer/mod.rs

```diff
@@ -0,0 +1,119 @@
+/// An sBTC signer configuration
+pub mod config;
+
+use std::collections::HashMap;
+
+use crate::signer::config::Config;
+use crate::SBTCError;
+use bitcoin::{Network, PrivateKey, PublicKey, Transaction as BitcoinTransaction};
+use p256k1::ecdsa;
+use url::Url;
+use wsts::{bip340::SchnorrProof, common::Signature};
+
+#[derive(Default, Clone, Debug)]
+pub struct SignersData {
+    pub signers: HashMap<u32, ecdsa::PublicKey>,
+    pub vote_shares: HashMap<u32, ecdsa::PublicKey>,
+}
+
+/// Coordinator algorithm trait
+pub trait CoordinatorAlgorithm {
+    fn generate_sbtc_wallet_public_key(&self) -> Result<PublicKey, SBTCError>;
+    fn run_signing_round(
+        &self,
+        _tx: BitcoinTransaction,
+    ) -> Result<(Signature, SchnorrProof), SBTCError>;
+}
+
+/// Signer algorithm trait
+pub trait SignerAlgorithm: CoordinatorAlgorithm {
+    fn sign_message(&self, message: &[u8]) -> Result<Vec<u8>, SBTCError>;
+    fn verify_message(
+        &self,
+        public_key: &ecdsa::PublicKey,
+        message: &[u8],
+    ) -> Result<bool, SBTCError>;
+}
+
+/// sBTC compliant Signer
+pub struct Signer<S: SignerAlgorithm> {
+    /// Signer configuration
+    pub config: Config,
+    /// Signer private key
+    pub private_key: PrivateKey,
+    /// Network to use
+    pub network: Network,
+    /// The stacks node RPC URL
+    pub stacks_node_rpc_url: Url,
+    /// The signing algorithm
+    pub algorithm: S,
+}
+
+impl<S: SignerAlgorithm> Signer<S> {
+    /// Create a new signer
+    pub fn new(
+        config: Config,
+        private_key: PrivateKey,
+        network: Network,
+        stacks_node_rpc_url: Url,
+        algorithm: S,
+    ) -> Self {
+        Self {
+            config,
+            private_key,
+            network,
+            stacks_node_rpc_url,
+            algorithm,
+        }
+    }
+
+    /// Get the current coordinator public key from the stacks node
+    pub fn get_coordinator_public_key(&self) -> Result<PublicKey, SBTCError> {
+        todo!()
+    }
+
+    /// Retrieve the current signers' data from the stacks node
+    pub fn get_signers_data(&self) -> Result<SignersData, SBTCError> {
+        todo!()
+    }
+
```

Comment:
Nit: I wouldn't specify how the data is obtained here, only require that the data is available. In practice I assume this data is not fetched from the node on every request, but rather stored in an internal database (or in memory) for the signer, which subscribes to updates from the node to keep this data in sync.

Reply:
Yeah, from experience getters should be implemented per trait implementor and then higher level functions can build default shared functionality on top of them.

Reply:
This would most likely be stored in the stacker db which is accessed via a stacks node.